internal/controller/apiextensions/definition/handlers.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package definition
+
+import (
+	"context"
+
+	kunstructured "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kevent "sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured/composite"
+
+	v1 "github.com/crossplane/crossplane/apis/apiextensions/v1"
+)
+
+// EnqueueForCompositionRevision enqueues reconciles for all XRs that will use a
+// newly created CompositionRevision.
+func EnqueueForCompositionRevision(of resource.CompositeKind, c client.Reader, log logging.Logger) handler.Funcs {
+	return handler.Funcs{
+		CreateFunc: func(ctx context.Context, e kevent.CreateEvent, q workqueue.RateLimitingInterface) {
+			rev, ok := e.Object.(*v1.CompositionRevision)
+			if !ok {
+				// should not happen
+				return
+			}
+
+			// TODO(negz): Check whether the revision's compositeTypeRef matches
+			// the supplied CompositeKind. If it doesn't, we can return early.
+
+			// get all XRs
+			xrs := kunstructured.UnstructuredList{}
+			xrs.SetGroupVersionKind(schema.GroupVersionKind(of))
+			xrs.SetKind(schema.GroupVersionKind(of).Kind + "List")
+			if err := c.List(ctx, &xrs); err != nil {
```

Comment:
Should we consider indexing with composition revisions to make this more efficient? Also wondering if we would need paging after some scale 🤔

Reply:
Possibly. Right now this logic is unchanged by this PR, so I've just added a TODO to address in future. 🙂